### PR TITLE
use discriminator on the backend to deserialise product object

### PR DESCRIPTION
**Pull request by johnduffell**
## What are you doing in this PR?

Following on from https://github.com/guardian/support-frontend/pull/2866

This adds the code on the server side to pay attention to the incoming productType field in the client side payload.

https://trello.com/c/Mwb7aMda/3379-add-discriminator-to-producttype

## Why are you doing this?

Same reasons as the above PR.

## Testing

All tests and IT tests working from support-frontend and support-workers
Need final test in CODE once approved

### CODE REVIEW

File: support-models/src/test/scala/com/gu/support/workers/SerialisationSpec.scala [hunk 1]
```diff
@@ -16,11 +16,6 @@ import org.scalatest.flatspec.AnyFlatSpec
 
 class SerialisationSpec extends AnyFlatSpec with SerialisationTestHelpers with LazyLogging with EitherValues {
 
-  "Contribution JSON with a billing period set" should "be decoded correctly" in {
-    val input = contribution(billingPeriod = Annual)
-    testDecoding[Contribution](input, _.billingPeriod shouldBe Annual)
-  }
```
Comment (johnduffell) [inline support-models/src/test/scala/com/gu/support/workers/SerialisationSpec.scala:22]: this is covered in the above test class

File: support-models/src/main/scala/com/gu/support/workers/Products.scala [hunk 2]
```diff
@@ -57,25 +57,14 @@ case class GuardianWeekly(
 
 object ProductType {
   import com.gu.support.encoding.CustomCodecs._
-  implicit val decoderDigital: Decoder[DigitalPack] = deriveDecoder[DigitalPack]
-    .prepare(_.withFocus(_.mapObject(_.checkKeyExists("readerType", Json.fromString("Direct")))))
```
Comment (johnduffell) [inline support-models/src/main/scala/com/gu/support/workers/Products.scala:61]: we used to accept a missing readerType, but it looks like the frontend does send it.  So I have dropped that functionality as the standard `DiscriminatedType` doesn't have facility for special cases at the moment.
It wouldn't be too hard to add the extra complexity, but I don't know if it's too big of a risk to drop that?  Thoughts please.

Comment (johnduffell) [reply support-models/src/main/scala/com/gu/support/workers/Products.scala:61]: this was added only in may for corporate subs, apparently just to get the tests working https://github.com/guardian/support-frontend/pull/2484/files#diff-6bc499707cdd7d347bfaa7ce1da6dcf4fc67d86af1c0ca30c801f4e3d1b704f1R64
and changed later to ReaderType 
https://github.com/guardian/support-frontend/pull/2635/files#r463085289

Comment (rupertbates) [reply support-models/src/main/scala/com/gu/support/workers/Products.scala:61]: If it is always sent, which would make sense, then I think we should remove it

**Review by rupertbates — APPROVED** [review]
LGTM

**Pull request merged**